controller/backing_image_manager_controller.go

```diff
@@ -0,0 +1,1233 @@
+			if bim.Name == currentBIM.Name {
+				continue
+			}
+			if bim.Status.CurrentState != types.BackingImageManagerStateRunning {
+				continue
+			}
+			info, exists := bim.Status.BackingImageFileMap[biName]
+			if !exists {
+				continue
+			}
+			if info.State == types.BackingImageDownloadStateFailed {
+				continue
+			}
+			pullRequired = false
+			// Use images in default manager as senders only
+			if bim.Spec.Image == defaultImage && info.State == types.BackingImageDownloadStateDownloaded && info.SendingReference < bimtypes.SendingLimit {
+				senderCandidate = bim
+				break
+			}
+		}
+
+		if pullRequired {
+			isEligible, err := c.isEligibleForPulling(currentBIM, biName)
+			if err != nil {
+				return err
+			}
+			if !isEligible {
+				log.Debugf("Current backing image manager is not eligible for pulling")
+				continue
+			}
+			log.Debugf("Start to pull backing image")
+			if _, err := cli.Pull(bi.Name, bi.Spec.ImageURL, bi.Status.UUID); err != nil {
+				if types.ErrorAlreadyExists(err) {
+					log.Debugf("Backing image already exists, no need to pull it again")
+					continue
+				}
+				return err
+			}
+			log.Debugf("Pulling backing image")
+			c.eventRecorder.Eventf(currentBIM, v1.EventTypeNormal, EventReasonPulling, "Pulling backing image %v in disk %v on node %v", bi.Name, currentBIM.Spec.DiskUUID, currentBIM.Spec.NodeID)
+			continue
+		}
+		if senderCandidate != nil {
+			log.WithFields(logrus.Fields{"fromHost": senderCandidate.Status.IP, "toHost": currentBIM.Status.IP, "size": bi.Status.Size}).Debugf("Start to sync backing image")
+			if _, err := cli.Sync(biName, bi.Spec.ImageURL, bi.Status.UUID, senderCandidate.Status.IP, currentBIM.Status.IP, bi.Status.Size); err != nil {
+				if types.ErrorAlreadyExists(err) {
+					log.WithFields(logrus.Fields{"fromHost": senderCandidate.Status.IP, "toHost": currentBIM.Status.IP, "size": bi.Status.Size}).Debugf("Backing image already exists, no need to sync from others")
+					continue
+				}
+				return err
+			}
+			log.WithFields(logrus.Fields{"fromHost": senderCandidate.Status.IP, "toHost": currentBIM.Status.IP, "size": bi.Status.Size}).Debugf("Syncing backing image")
+			c.eventRecorder.Eventf(currentBIM, v1.EventTypeNormal, EventReasonSyncing, "Syncing backing image %v in disk %v on node %v from %v(%v)", bi.Name, currentBIM.Spec.DiskUUID, currentBIM.Spec.NodeID, senderCandidate.Name, senderCandidate.Status.IP)
+			continue
+		}
+	}
+
+	return nil
+}
+
+func (c *BackingImageManagerController) isEligibleForPulling(currentBIM *longhorn.BackingImageManager, biName string) (bool, error) {
+	defaultBIMs, err := c.ds.ListDefaultBackingImageManagers()
+	if err != nil {
+		return false, err
+	}
+	readyCandidateNames := []string{}
+	for bimName, bim := range defaultBIMs {
+		if _, exists := bim.Spec.BackingImages[biName]; !exists {
+			continue
+		}
+		isDownOrDeleted, err := c.ds.IsNodeDownOrDeleted(bim.Spec.NodeID)
+		if err != nil {
+			return false, err
+		}
```

Comment:
Thinking about this case:
When the 1st backing image is created in the Longhorn system, there is no BIM. To download the backing image, the backing image controller will directly create BIMs for all related nodes. Then all related backing image manager controllers will create the pods and update the state of BIMs. For each backing image manager controller, it may consider the current BIM is the only running one (state `running` may have not been recorded in ETCD), but actually all/most of BIMs will become running simultaneously. In this case, all BIMs will start pulling. 

That's why I give up checking `bim.Status.CurrentState == types.BackingImageManagerStateRunning` here. With the current implementation, we cannot eliminate the above scenario but it reduces the possibility of multiple pulling.

Reply:
I see the reason.

But we should at least ignore the error BIM, do we? if the preferred candidate somehow cannot become running, it probably will be marked error, and we can safely ignore it.

Reply:
Agree. Longhorn can tolerate occasional duplicate downloading but cannot accept any blocking. And it needs to check the disk readiness as well.